Classes/Routing/PageRouter.php

```diff
@@ -0,0 +1,215 @@
+        unset($parameters['_frontendHost']);
+        if ($languageOption instanceof SiteLanguage) {
+            $language = $languageOption;
+        } elseif ($languageOption !== null) {
+            $language = $this->site->getLanguageById((int)$languageOption);
+        }
+        if ($language === null) {
+            $language = $this->site->getDefaultLanguage();
+        }
+
+        $pageId = 0;
+        if (is_array($route)) {
+            $pageId = (int)$route['uid'];
+        } elseif (is_scalar($route)) {
+            $pageId = (int)$route;
+        }
+
+        $context = clone $this->context;
+        $context->setAspect('language', LanguageAspectFactory::createFromSiteLanguage($language));
+        $pageRepository = GeneralUtility::makeInstance(PageRepository::class, $context);
+        $page = $pageRepository->getPage($pageId, true);
+        $pagePath = $page['slug'] ?? '';
+
+        if ($parameters['MP'] ?? false) {
+            $mountPointPairs = explode(',', $parameters['MP']);
+            $pagePath = $this->resolveMountPointParameterIntoPageSlug(
+                $pageId,
+                $pagePath,
+                $mountPointPairs,
+                $pageRepository
+            );
+
+            // If the MountPoint page has a different site, the link needs to be generated
+            // with the base of the MountPoint page, this is especially relevant for cross-domain linking
+            // Because the language contains the full base, it is retrieved in this case.
+            try {
+                [, $mountPointPage] = explode('-', reset($mountPointPairs));
+                $site = GeneralUtility::makeInstance(SiteMatcher::class)
+                    ->matchByPageId((int)$mountPointPage);
+                $language = $site->getLanguageById($language->getLanguageId());
+            } catch (SiteNotFoundException $e) {
+                // No alternative site found, use the existing one
+            }
+            // Store the MP parameter in the page record, so it could be used for any enhancers
+            $page['MPvar'] = $parameters['MP'];
+            unset($parameters['MP']);
+        }
+
+        $originalParameters = $parameters;
+        $collection = new RouteCollection();
+        $defaultRouteForPage = new Route(
+            '/' . ltrim($pagePath, '/'),
+            [],
+            [],
+            ['utf8' => true, '_page' => $page]
+        );
+        $collection->add('default', $defaultRouteForPage);
+
+        // cHash is never considered because cHash is built by this very method.
+        unset($originalParameters['cHash']);
+        $enhancers = $this->getEnhancersForPage($pageId, $language);
+        foreach ($enhancers as $enhancer) {
+            if ($enhancer instanceof RoutingEnhancerInterface) {
+                $enhancer->enhanceForGeneration($collection, $originalParameters);
+            }
+        }
+        foreach ($enhancers as $enhancer) {
```

Comment:
You can use one loop with switch for enhacers

Reply:
This is Xclass for existing core method, we don't want to refractor this code, only add new feature